File: app/src/main/java/com/dhpcs/liquidity/fragment/TransfersFragment.java
package com.dhpcs.liquidity.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.util.SortedList;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.util.SortedListAdapterCallback;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.dhpcs.liquidity.BoardGame.Player;
import com.dhpcs.liquidity.BoardGame.TransferWithCurrency;
import com.dhpcs.liquidity.R;
import com.dhpcs.liquidity.activity.BoardGameActivity;

import java.text.DateFormat;
import java.util.ArrayList;
import java.util.List;

import scala.collection.Iterator;

public class TransfersFragment extends Fragment {

    private class TransfersAdapter
            extends RecyclerView.Adapter<TransfersAdapter.TransferViewHolder> {

        public class TransferViewHolder extends RecyclerView.ViewHolder {

            private final TextView textViewSummary;
            private final TextView textViewCreatedTime;
            private final TextView textViewCreatedDate;

            public TransferViewHolder(View itemView) {
                super(itemView);
                textViewSummary = (TextView) itemView.findViewById(R.id.textview_summary);
                textViewCreatedTime = (TextView) itemView.findViewById(R.id.textview_created_time);
                textViewCreatedDate = (TextView) itemView.findViewById(R.id.textview_created_date);
            }

            public void bindTransfer(TransferWithCurrency transfer) {

                boolean isFromPlayer = player != null && transfer.from().right().get().member().id()
                        .equals(player.member().id());
                String value = BoardGameActivity.formatCurrencyValue(
                        context,
                        transfer.currency(),
                        transfer.transaction().value()
                );
                boolean isToPlayer = player != null && transfer.to().right().get().member().id()
                        .equals(player.member().id());
                String summary;
                if (isFromPlayer && !isToPlayer) {
                    summary = context.getString(
                            R.string.transfer_summary_sent_to_format_string,
                            value,
                            BoardGameActivity.formatMemberOrAccount(context, transfer.to())
                    );
                } else if (!isFromPlayer && isToPlayer) {
                    summary = context.getString(
                            R.string.transfer_summary_received_from_format_string,
                            value,
                            BoardGameActivity.formatMemberOrAccount(context, transfer.from())
                    );
                } else {
                    summary = context.getString(
                            R.string.transfer_summary_format_string,
                            BoardGameActivity.formatMemberOrAccount(context, transfer.from()),
                            value,
                            BoardGameActivity.formatMemberOrAccount(context, transfer.to())
                    );
                }
                String createdTime = getString(
                        R.string.transfer_created_time_format_string,
                        timeFormat.format(transfer.transaction().created())
                );
                String createdDate = getString(
                        R.string.transfer_created_date_format_string,
                        dateFormat.format(transfer.transaction().created())
                );

                textViewSummary.setText(summary);
                textViewCreatedTime.setText(createdTime);
                textViewCreatedDate.setText(createdDate);
            }

        }

        private final Context context;
        private final Player player;
        private final DateFormat timeFormat = DateFormat.getTimeInstance();
        private final DateFormat dateFormat = DateFormat.getDateInstance();
        private final SortedList<TransferWithCurrency> transfers = new SortedList<>(
                TransferWithCurrency.class,
                new SortedListAdapterCallback<TransferWithCurrency>(this) {

                    @Override
                    public int compare(TransferWithCurrency o1,
                                       TransferWithCurrency o2) {
                        long lhsId = o1.transaction().id().id();
                        long rhsId = o2.transaction().id().id();
                        return -1 *
                                (lhsId < rhsId ? -1 : (lhsId == rhsId ? 0 : 1));
                    }

                    @Override
                    public boolean areContentsTheSame(TransferWithCurrency oldItem,
                                                      TransferWithCurrency newItem) {
                        return oldItem.equals(newItem);
                    }

                    @Override
                    public boolean areItemsTheSame(TransferWithCurrency item1,
                                                   TransferWithCurrency item2) {
                        return item1.transaction().id().equals(item2.transaction().id());
                    }

                }
        );

        public TransfersAdapter(Context context, Player player) {
            this.context = context;
            this.player = player;
        }

        public void beginBatchedUpdates() {
            transfers.beginBatchedUpdates();
        }

        public void endBatchedUpdates() {
            transfers.endBatchedUpdates();
        }

        public void replaceOrAdd(TransferWithCurrency transfer) {
            transfers.add(transfer);
        }

        @Override
        public int getItemCount() {
            return transfers.size();
        }

        @Override
        public TransferViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = LayoutInflater
                    .from(parent.getContext())
                    .inflate(R.layout.linearlayout_transfer, parent, false);
            return new TransferViewHolder(view);
        }

        @Override
        public void onBindViewHolder(TransferViewHolder holder, int position) {
            TransferWithCurrency transfer = transfers.get(position);
            holder.bindTransfer(transfer);
        }

    }

    private static final String ARG_PLAYER = "player";
    private static final String ARG_TRANSFERS = "transfers";

    public static TransfersFragment newInstance(Player player,
                                                ArrayList<TransferWithCurrency> transfers) {
        TransfersFragment transfersFragment = new TransfersFragment();
        Bundle args = new Bundle();
        args.putSerializable(ARG_PLAYER, player);
        args.putSerializable(ARG_TRANSFERS, transfers);
        transfersFragment.setArguments(args);
        return transfersFragment;
    }

    private TransfersAdapter transfersAdapter;

    private TextView textViewEmpty;
    private RecyclerView recyclerViewTransfers;

    public void onTransferAdded(TransferWithCurrency addedTransfer) {
        replaceOrAddTransfer(
                (Player) getArguments().getSerializable(ARG_PLAYER),
                addedTransfer
        );
        textViewEmpty.setVisibility(View.GONE);
        recyclerViewTransfers.setVisibility(View.VISIBLE);
    }

    public void onTransfersChanged(
            scala.collection.Iterable<TransferWithCurrency> changedTransfers) {
        replaceOrAddTransfers(changedTransfers);
    }

    public void onTransfersInitialized(
            scala.collection.Iterable<TransferWithCurrency> transfers) {
        replaceOrAddTransfers(transfers);
        if (transfers.nonEmpty()) {
            textViewEmpty.setVisibility(View.GONE);
            recyclerViewTransfers.setVisibility(View.VISIBLE);
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Player player = (Player) getArguments().getSerializable(ARG_PLAYER);
        List<TransferWithCurrency> transfers =
                (ArrayList<TransferWithCurrency>) getArguments().getSerializable(ARG_TRANSFERS);

        transfersAdapter = new TransfersAdapter(getActivity(), player);

        if (transfers != null) {
            transfersAdapter.beginBatchedUpdates();
            for (TransferWithCurrency transfer : transfers) {
                replaceOrAddTransfer(player, transfer);
            }
            transfersAdapter.endBatchedUpdates();
        }

    }

    @Override
    public View onCreateView(LayoutInflater inflater,
                             ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_transfers, container, false);

        textViewEmpty = (TextView) view.findViewById(R.id.textview_empty);
        recyclerViewTransfers = (RecyclerView) view.findViewById(R.id.recyclerview_transfers);

        recyclerViewTransfers.setHasFixedSize(true);
        recyclerViewTransfers.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerViewTransfers.setAdapter(transfersAdapter);

        if (transfersAdapter.getItemCount() != 0) {
            textViewEmpty.setVisibility(View.GONE);
            recyclerViewTransfers.setVisibility(View.VISIBLE);
        }

        return view;
    }

    private void replaceOrAddTransfers(scala.collection.Iterable<TransferWithCurrency> transfers) {
        Player player = (Player) getArguments().getSerializable(ARG_PLAYER);
        Iterator<TransferWithCurrency> iterator = transfers.iterator();
        while (iterator.hasNext()) {
            replaceOrAddTransfer(player, iterator.next());
        }
    }

    private void replaceOrAddTransfer(Player player, TransferWithCurrency transfer) {
        if (player == null || (transfer.from().isRight()
                && player.member().id().equals(transfer.from().right().get().member().id()))
                || (transfer.to().isRight()
                && player.member().id().equals(transfer.to().right().get().member().id()))) {
            transfersAdapter.replaceOrAdd(transfer);
        }
    }

}

File: app/src/main/java/com/dhpcs/liquidity/fragment/TransfersFragment.kt
package com.dhpcs.liquidity.fragment

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.*
import com.dhpcs.liquidity.BoardGame
import com.dhpcs.liquidity.LiquidityApplication
import com.dhpcs.liquidity.R
import com.dhpcs.liquidity.activity.MainActivity
import com.dhpcs.liquidity.activity.MainActivity.Companion.liveData
import kotlinx.android.synthetic.main.fragment_transfers.*
import java.text.DateFormat

class TransfersFragment : Fragment() {

    companion object {

        private const val ARG_PLAYER_ID = "player_id"

        private val timeFormat = DateFormat.getTimeInstance()
        private val dateFormat = DateFormat.getDateInstance()

        fun newInstance(player: BoardGame.Companion.Player?): TransfersFragment {
            val transfersFragment = TransfersFragment()
            val args = Bundle()
            args.putString(ARG_PLAYER_ID, player?.memberId)
            transfersFragment.arguments = args
            return transfersFragment
        }

        private class TransfersAdapter
        internal constructor(private val player: BoardGame.Companion.Player?) :
                ListAdapter<BoardGame.Companion.Transfer, TransferViewHolder>(
                        object : DiffUtil.ItemCallback<BoardGame.Companion.Transfer>() {

                            override fun areContentsTheSame(
                                    oldItem: BoardGame.Companion.Transfer,
                                    newItem: BoardGame.Companion.Transfer
                            ): Boolean {
                                val isFromPlayer = player != null &&
                                        oldItem.fromPlayer?.memberId == player.memberId
                                val isToPlayer = player != null &&
                                        oldItem.toPlayer?.memberId == player.memberId
                                return if (isFromPlayer && !isToPlayer) {
                                    oldItem.toPlayer?.name == newItem.toPlayer?.name
                                } else if (!isFromPlayer && isToPlayer) {
                                    oldItem.fromPlayer?.name == newItem.fromPlayer?.name
                                } else {
                                    oldItem.fromPlayer?.name == newItem.fromPlayer?.name &&
                                            oldItem.toPlayer?.name == newItem.toPlayer?.name
                                }
                            }

                            override fun areItemsTheSame(
                                    item1: BoardGame.Companion.Transfer,
                                    item2: BoardGame.Companion.Transfer
                            ): Boolean = item1.transactionId == item2.transactionId

                        }
                ) {

            private val transferComparator =
                    Comparator<BoardGame.Companion.Transfer> { o1, o2 ->
                        val lhsCreated = o1.created
                        val rhsCreated = o2.created
                        when {
                            lhsCreated < rhsCreated -> 1
                            lhsCreated > rhsCreated -> -1
                            else -> {
                                val lhsId = o1.transactionId
                                val rhsId = o2.transactionId
                                lhsId.compareTo(rhsId)
                            }
                        }
                    }

            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TransferViewHolder {
                val view = LayoutInflater
                        .from(parent.context)
                        .inflate(R.layout.linearlayout_transfer, parent, false)
                return TransferViewHolder(view, parent.context, player)
            }

            override fun onBindViewHolder(holder: TransferViewHolder, position: Int) {
                holder.bindTransfer(getItem(position))
            }

            internal fun updateTransfers(
                    transfers: Collection<BoardGame.Companion.Transfer>) {
                submitList(transfers.sortedWith(transferComparator))
            }

        }

        private class TransferViewHolder(itemView: View,
                                         private val context: Context,
                                         private val player: BoardGame.Companion.Player?
        ) : RecyclerView.ViewHolder(itemView) {

            private val textViewSummary =
                    itemView.findViewById<TextView>(R.id.textview_summary)
            private val textViewCreatedTime =
                    itemView.findViewById<TextView>(R.id.textview_created_time)
            private val textViewCreatedDate =
                    itemView.findViewById<TextView>(R.id.textview_created_date)

            fun bindTransfer(transfer: BoardGame.Companion.Transfer) {
                val isFromPlayer = player != null &&
                        transfer.fromPlayer?.memberId == player.memberId
                val value = LiquidityApplication.formatCurrencyValue(
                        context,
                        transfer.currency,
                        transfer.value
                )
                val isToPlayer = player != null &&
                        transfer.toPlayer?.memberId == player.memberId
                val summary = if (isFromPlayer && !isToPlayer) {
                    context.getString(
                            R.string.transfer_summary_sent_to_format_string,
                            value,
                            LiquidityApplication.formatNullable(context, transfer.toPlayer?.name)
                    )
                } else if (!isFromPlayer && isToPlayer) {
                    context.getString(
                            R.string.transfer_summary_received_from_format_string,
                            value,
                            LiquidityApplication.formatNullable(context, transfer.fromPlayer?.name)
                    )
                } else {
                    context.getString(
                            R.string.transfer_summary_format_string,
                            LiquidityApplication.formatNullable(context, transfer.fromPlayer?.name),
                            value,
                            LiquidityApplication.formatNullable(context, transfer.toPlayer?.name)
                    )
                }
                val createdTime = context.getString(
                        R.string.transfer_created_time_format_string,
                        timeFormat.format(transfer.created)
                )
                val createdDate = context.getString(
                        R.string.transfer_created_date_format_string,
                        dateFormat.format(transfer.created)
                )
                textViewSummary.text = summary
                textViewCreatedTime.text = createdTime
                textViewCreatedDate.text = createdDate
            }

        }

    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_transfers, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        val model = ViewModelProviders.of(requireActivity())
                .get(MainActivity.Companion.BoardGameModel::class.java)

        val playerId = arguments!!.getString(ARG_PLAYER_ID)

        val player = if (playerId != null) {
            model.boardGame.players[playerId]
        } else {
            null
        }

        val transfersAdapter = TransfersAdapter(player)

        recyclerview_transfers.setHasFixedSize(true)
        recyclerview_transfers.layoutManager = LinearLayoutManager(activity)
        recyclerview_transfers.adapter = transfersAdapter
        recyclerview_transfers.addItemDecoration(DividerItemDecoration(
                recyclerview_transfers.context,
                DividerItemDecoration.VERTICAL
        ))

        if (transfersAdapter.itemCount != 0) {
            textview_empty.visibility = View.GONE
            recyclerview_transfers.visibility = View.VISIBLE
        }

        model.boardGame.liveData { it.transfersObservable }.observe(this, Observer {
            val visibleTransfers = it.filter { transfer ->
                player == null ||
                        (transfer.fromPlayer != null &&
                                player.memberId == transfer.fromPlayer.memberId) ||
                        (transfer.toPlayer != null &&
                                player.memberId == transfer.toPlayer.memberId)
            }
            transfersAdapter.updateTransfers(visibleTransfers)
            if (visibleTransfers.isNotEmpty()) {
                textview_empty.visibility = View.GONE
                recyclerview_transfers.visibility = View.VISIBLE
            } else {
                textview_empty.visibility = View.VISIBLE
                recyclerview_transfers.visibility = View.GONE
            }
        })
    }

}
